rapid7_insightidr/komand_rapid7_insightidr/actions/advanced_query/action.py

```diff
@@ -0,0 +1,191 @@
+
+        log_entries = komand.helper.clean(log_entries)
+
+        for log_entry in log_entries:
+            log_entry["message"] = json.loads(log_entry.get("message", "{}"))
+
+        self.logger.info(f"Sending results to orchestrator.")
+        return {Output.RESULTS: log_entries}
+
+    def parse_dates(self, time_from_string: str, time_to_string: str) -> (int, int):
+        """
+        Parse incoming dates and return them as millisecond epoch time
+
+        @param time_from_string: str
+        @param time_to_string: str (optional, if it's a falsey value, time to will be set to Now)
+        @return: (int, int)
+        """
+
+        # Parse times to epoch milliseconds
+        try:
+            time_from = int(parse(time_from_string).timestamp()) * 1000
+            if time_to_string:
+                time_to = int(parse(time_to_string).timestamp()) * 1000
+            else:
+                # Now in millisecond epoch time
+                time_to = int(time.time()) * 1000
+        except ParserError as e:
+            raise PluginException(cause="Could not parse given date.",
+                                  assistance="The date given was in an unrecognizable format.",
+                                  data=e)
+        return time_from, time_to
+
+    def get_results_from_callback(self, callback_url: str, timeout: int) -> [object]:
+        """
+        Get log entries from a callback URL
+
+        @param callback_url: str
+        @return: list of log entries
+        """
+        self.logger.info(f"Trying to get results from callback URL: {callback_url}")
+        response = self.connection.session.get(callback_url)
+        try:
+            response.raise_for_status()
+        except Exception:
+            raise PluginException(cause="Failed to get logs from InsightIDR",
+                                  assistance=f"Could not get logs from: {callback_url}",
+                                  data=response.text)
+        results_object = response.json()
+        log_entries = results_object.get("events")
+
+        counter = timeout
+        while not log_entries and counter > 0:
+            self.logger.info("Results were not ready. Sleeping 1 second and trying again.")
+            self.logger.info(f"Time left: {counter}")
+            time.sleep(1)
+            response = self.connection.session.get(callback_url)
+
+            # TODO: need to know what happens if we poll for a query that's not done
+            try:
+                response.raise_for_status()
+            except Exception:
+                raise PluginException(cause="Failed to get logs from InsightIDR",
+                                      assistance=f"Could not get logs from: {callback_url}",
+                                      data=response.text)
+
+            results_object = response.json()
+            log_entries = results_object.get("events")
+            counter-=1
+
```

Comment:
Is it worth it to raise a plugin exception (or just more specific logging) here if `log_entries` is empty by the time counter hits 0?

Reply:
Fixed, moved some code around.